### PR TITLE
Implement exponential backoff for retries

### PR DESCRIPTION
**Pull request by sonalidbhavsar**
Implement exponential backoff for retries #21

@alovak @adamdecaf

### CODE REVIEW

File: pool.go [hunk 1]
```diff
@@ -187,7 +188,8 @@ func (p *Pool) recreateConnection(closedConn *Connection) {
 
 		p.handleError(fmt.Errorf("failed to reconnect to %s: %w", conn.addr, err))
 		select {
-		case <-time.After(p.Opts.ReconnectWait):
+		case <-time.After(reconnectTime):
+			reconnectTime *= 2
```
Comment (adamdecaf) [inline pool.go:194]: Can we have an upper bound here? We don't want this to forever increase.

Comment (sonalidbhavsar) [reply pool.go:194]: Added MaxReconnectWait param, once MaxReconnectWait is reached, retry will stop.

File: pool.go [hunk 2]
```diff
@@ -186,8 +186,14 @@ func (p *Pool) recreateConnection(closedConn *Connection) {
 		}
 
 		p.handleError(fmt.Errorf("failed to reconnect to %s: %w", conn.addr, err))
+
 		select {
-		case <-time.After(p.Opts.ReconnectWait):
+		case <-time.After(reconnectTime):
+			reconnectTime *= 2
+			if reconnectTime > p.Opts.MaxReconnectWait {
+				// stop connection retry after MaxReconnectWait
```
Comment (alovak) [inline pool.go]: we can continue to set `reconnectTime` to `MaxReconnectTime` without returning from the re-try loop.

**Comment by alovak** [comment]
@sonalidbhavsar, thanks for your contributions! I like the change. it's small and straightforward and no new dependencies.

@adamdecaf I was thinking about https://github.com/sethvargo/go-retry/blob/main/backoff_exponential.go, but I'm not sure what benefits it may give us. What do you think about the approach in this PR?

**Comment by adamdecaf** [comment]
> @adamdecaf I was thinking about https://github.com/sethvargo/go-retry/blob/main/backoff_exponential.go, but I'm not sure what benefits it may give us. What do you think about the approach in this PR?

We use that library elsewhere in moov-io code. It's nice, but if you want to avoid adding dependencies we could do the math here.

File: pool.go (hunk 2 above)
Comment (sonalidbhavsar) [reply pool.go]: As I understand, we don't want to stop connection retry ever, once we reach  MaxReconnectTime, we can stop increasing reconnectTime, and keep retrying every MaxReconnectTime duration.

The changes as per this are done.

**Comment by sonalidbhavsar** [comment]
> > @adamdecaf I was thinking about https://github.com/sethvargo/go-retry/blob/main/backoff_exponential.go, but I'm not sure what benefits it may give us. What do you think about the approach in this PR?
> 
> We use that library elsewhere in moov-io code. It's nice, but if you want to avoid adding dependencies we could do the math here.

I have implemented the alternative approach, using the mentioned library. You can check it here: https://github.com/sonalidbhavsar/iso8583-connection/pull/2/files

I can raise PR with alternative approach, if required.

**Comment by adamdecaf** [comment]
ping @alovak

**Comment by alovak** [comment]
@sonalidbhavsar thanks a lot for the PR! I've used it as a base for #66 where I merged with the master and did some code reorg. I apologize for such a long delay in merging it.

**Pull request closed without merging**